### PR TITLE
fix(terminal): Terminal::insert_before would crash when called while the viewport filled the screen

### PR DESCRIPTION
**Pull request by nfachan**
Reimplement Terminal::insert_before. The previous implementation would insert the new lines in chunks into the area between the top of the screen and the top of the (new) viewport. If the viewport filled the screen, there would be no area in which to insert lines, and the function would crash.

The new implementation uses as much of the screen as it needs to, all the way up to using the whole screen.

This commit:
- adds a scrollback buffer to the `TestBackend` so that tests can inspect and assert the state of the scrollback buffer in addition to the screen
- adds functions to `TestBackend` to assert the state of the scrollback
- adds and updates `TestBackend` tests to test the behavior of the scrollback and the new asserting functions
- reimplements `Terminal::insert_before`, including adding two new helper functions `Terminal::draw_lines` and `Terminal::scroll_up`.
- updates the documentation for `Terminal::insert_before` to clarify some of the edge cases
- updates terminal tests to assert the state of the scrollback buffer
- adds a new test for the condition that causes the bug
- adds a conversion constructor `Cell::from(char)`

Fixes: https://github.com/ratatui/ratatui/issues/999

### CODE REVIEW

**Comment by orhun** [comment]
Would be nice to have unit tests for this to verify the behavior :)

**Comment by nfachan** [comment]
There are the existing tests in tests/terminal.rs that validate the previous behavior.

I have some commits I'm working on to add scrollback to the test backend so that I can enhance the existing tests to validate scrollback as well as add new tests for the case when the viewport covers the whole screen.

I was planning on putting those commits in a separate PR, but I can add them to this PR if you like.

**Comment by nfachan** [comment]
Okay, I've added some tests. Are there any more tests you would like to see?

File: src/backend/test.rs [hunk 1]
```diff
@@ -102,6 +112,38 @@ impl TestBackend {
...
+        // TODO: use assert_eq!()
+        crate::assert_buffer_eq!(&self.scrollback, expected);
+    }
+
+    /// Asserts that the `TestBackend`'s scrollback buffer is empty, but also has the given width.
+    ///
+    /// # Panics
+    /// When the scrollback buffer is not equal, a panic occurs with a detailed error message
+    /// showing the differences between the expected and actual buffers.
+    pub fn assert_empty_scrollback(&self, width: u16) {
```
Comment (orhun) [inline src/backend/test.rs:134]: I don't find the arguments to this function intuitive. Maybe it would be leave testing this to the user instead of adding a opinionated function like this?

**Comment by orhun** [comment]
Yup, it looks good now.

File: src/terminal/terminal.rs [hunk 2]
```diff
@@ -545,51 +580,162 @@ where
...
+
+        // Use i32 variables so we don't have worry about overflowed u16s when adding, or about
+        // negative results when subtracting.
+        let mut drawn_height: i32 = self.viewport_area.top().into();
+        let mut buffer_height: i32 = height.into();
+        let viewport_height: i32 = self.viewport_area.height.into();
+        let screen_height: i32 = self.last_known_area.height.into();
+
+        while buffer_height + viewport_height > screen_height {
+            // There isn't enough room on the screen for the whole buffer and the viewport. We will
```
Comment (orhun) [inline src/terminal/terminal.rs:603]: I appreciate this very detailed and explanatory comment. However, it feels a bit too much, should we strip some parts or simplify it maybe? Any thoughts @joshka @kdheepak?

File: src/backend/test.rs (hunk 1 above)
Comment (nfachan) [reply src/backend/test.rs:134]: The problem is that it's kind of a pain to specify an empty buffer using what's currently there. The assert is used in three places now, and I can imagine it being used in more in the future. Having to specify an empty buffer of the correct width at the assert-side is pretty clunky and verbose.

What about if I removed `width`? From the asserter's perspective, it doesn't really matter what the width is. It's really an implementation decision that the scrollback buffer starts out with any width at all. I think that may address your concern?

File: src/terminal/terminal.rs (hunk 2 above)
Comment (nfachan) [reply src/terminal/terminal.rs:603]: It is quite the wall of text. I'm happy to remove some or all of it.

It took me a bit of time to figure out what was there before, so I was hoping to save somebody in the future a similar problem when trying to figure out the magic formula I left behind. But maybe the comment is so long and hard to follow, that it's as hard or harder to follow than it would be to just try to figure out the magic formula.

Just let me know what you want me to do with the comment.

Comment (nfachan) [reply src/terminal/terminal.rs:603]: One approach could be to take out the proof. The comment could just end after the four constraints and the explanation of why they're important. An interested reader could then prove to themselves that the chosen `scroll_up` satisfied all four constraints, without being walked through it step-by-step.

File: src/backend/test.rs (hunk 1 above)
Comment (orhun) [reply src/backend/test.rs]: Makes sense / looks good now! Thanks :)

File: src/terminal/terminal.rs (hunk 2 above)
Comment (orhun) [reply src/terminal/terminal.rs:603]: Yes, that makes sense to me. You can simply add the proof to the related issue or this PR and link it in the comment for those who wants to read more about it.

**Comment by nfachan** [comment]
Here is a proof that has been removed from the code in `Terminal::insert_before` because of length.

```
}
Given:
    to_draw = min(buffer_height, screen_height)
and:
    scroll_up = max(0, drawn_height + to_draw - screen_height)

We want to prove that `scroll_up` satisfis four different constraints:
    (1) scroll_up >= 0
    (2) scroll_up <= drawn_height
    (3) drawn_height - scroll_up + to_draw <= screen_height
    or, equivalently:
        scroll_up >= drawn_height + to_draw - screen_height
    (4) drawn_height - scroll_up + to_draw >= screen_height - viewport_height
    or, equivalently:
        scroll_up <= drawn_height + to_draw - screen_height + viewport_height

(1) says that we don't want to scroll down. That should never be necessary when
inserting data before the inline viewport.

(2) says that we shouldn't scroll off more data than is currently on the screen. If
we were to do this, we'd end up with blank lines in the scrollback.

(3) says that after we scroll, we need to have enough room to draw `to_draw` lines
on the screen.

(4) says that we don't want to end drawing with a gap of more than `viewport_height`
size at the bottom of the screen. If we didn't have this constraint, we could end up
with the viewport higher on the screen after this function than before it was
called.

In order to prove that our choice of `scroll_up` satisfies these constraints, we
make use of two facts:
    (A) X <= Z and Y <= Z implies max(X, Y) <= Z for any X, Y, Z
    (B) X >= Z or Y >= Z implies max(X, Y) >= Z for any X, Y, Z
In our case, since scroll_up = max(0, drawn_height + to_draw - screen_height), we
use these variable assignments:
    X = 0
    Y = drawn_height + to_draw - screen_height

(1) is satisfied because (using fact (B)):
    X = 0 >= 0

(2) is satisfied because (using fact (A)):
    X = 0 <= drawn_height
    and
    Y = drawn_height + to_draw - screen_height
    where
    to_draw = min(buffer_height, screen_height)
        <= screen_height
    combined, implies
    Y <= drawn_height + screen_height - screen_height
        = drawn_height

(3) is satisfied because (using fact (B)):
    Y = drawn_height + to_draw - screen_height
        >= drawn_height + to_draw - screen_height

(4) is satistfied because (using fact (A)):
    First we show for X:
        Either (a) buffer_height >= screen_height or
        (b) buffer_height < screen_height.
        If (a):
            to_draw = min(buffer_height, screen_height) = screen_height
            implies
            drawn_height + to_draw - screen_height + viewport_height
                = drawn_height + screen_height - screen_height + viewport_height
                = drawn_height + viewport_height >= 0 = X
        If (b):
            to_draw = min(buffer_height, screen_height) = buffer_height
        And the loop invariant tells us:
            buffer_height + viewport_height > screen_height
        So
            drawn_height + to_draw - screen_height + viewport_height
                = drawn_height + buffer_height - screen_height + viewport_height
                = drawn_height - screen_height + buffer_height + viewport_height
                > drawn_height - screen_height + screen_height
                = drawn_height
                >= 0 = X
    Now we show for Y:
        Y = drawn_height + to_draw - screen_height
            <= drawn_height + to_draw - screen_height + viewport_height       
```

File: src/terminal/terminal.rs (hunk 2 above)
Comment (nfachan) [reply src/terminal/terminal.rs:603]: I have moved the proof to a comment in this PR, and changed the code's comment accordingly.

File: src/terminal/terminal.rs [hunk 3]
```diff
@@ -493,8 +495,41 @@ where
         self.backend.size()
     }
 
+    /// Draw lines at the given vertical offset. The slice of cells must contain enough cells
+    /// for the requested lines. A slice of the unused cells are returned.
+    fn draw_lines<'a>(
```
Comment (joshka) [inline src/terminal/terminal.rs:500]: Move this below where it's called from

File: src/terminal/terminal.rs (hunk 2 above)
Comment (joshka) [reply src/terminal/terminal.rs:603]: By moving around the code so that the top level method appears before the code which it calls, you might find that the detail you have here naturally gravitates into the doc comments of the private functions that you're calling instead of being in a single place here.

The main problem I see with this comment and the one above this is that it dives into the detail a bit too fast without setting up the context. This is likely a good amount of detail about things you found out while implementing, but without having gone through that process with you it seems like it's not tethered to an understanding of what this code is supposed to be doing. Starting with the assumption that the reader of this knows nothing about what's going on here (this is reasonably close to correct for me at least ;)), try to explain the algorithm / concerns from the outside in rather than the inside out.

File: src/terminal/terminal.rs [hunk 4]
```diff
@@ -545,51 +580,108 @@ where
...
-                .viewport_area
-                .y
-                .saturating_add(height)
-                .min(viewport_at_bottom),
-            ..self.viewport_area
-        });
-
-        // Draw contents into buffer
+        // Draw contents to insert into a buffer. The rest of this function is concerned with
+        // drawing this buffer onto the screen.
```
Comment (joshka) [inline src/terminal/terminal.rs:584]: This seems like an odd comment that doesn't help me understand what's happening.

File: src/terminal/terminal.rs [hunk 5]
```diff
@@ -545,51 +580,108 @@ where
...
-                    c,
-                )
-            });
-            self.backend.draw(iter)?;
-            self.backend.flush()?;
-            self.set_cursor_position(self.viewport_area.as_position())?;
+        let mut buffer = buffer.content.as_slice();
+
+        // Use i32 variables so we don't have worry about overflowed u16s when adding, or about
+        // negative results when subtracting.
```
Comment (joshka) [inline src/terminal/terminal.rs:596]: Does it make sense to instead use saturating_add / sub methods for these, or are there intermediate points where something may go negative and then go back positive which would not work?

**Review by joshka — CHANGES_REQUESTED** [review]
I haven't reviewed this in its entirety - especially not a lot of the terminal code, as I think there's some things which would help clear up how this works that would require a significant re-review

**Comment by joshka** [comment]
> Here is a proof that has been removed from the code in `Terminal::insert_before` because of length.

IMO, The right place to put this sort of info is in unit tests. There's likely a rephrasing of a bunch of those statements along the lines of "Given blah, when blah, then blah" each of which can be translated into a simple test that shows this is true (for the test backend at least).

**Comment by joshka** [comment]
It could be worth splitting the test backend changes from the terminal changes to help make it easier to review this PR. @orhun what do you think about that idea?

File: src/backend/test.rs [hunk 6]
```diff
@@ -83,9 +85,17 @@ impl TestBackend {
         &self.buffer
     }
 
+    /// Returns a reference to the internal scrollback buffer of the `TestBackend`.
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:88]: Do you mean what the scrollback is?

File: src/backend/test.rs [hunk 7]
```diff
@@ -102,6 +112,38 @@ impl TestBackend {
...
+            area: Rect {
+                x: 0,
+                y: 0,
+                width: self.scrollback.area.width,
+                height: 0,
+            },
+            content: vec![],
+        };
+        self.assert_scrollback(&expected);
+    }
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:145]: I have a change that renames the function as you suggested. There is already `assert_scrollback_lines`, so it only makes sense to have `assert_scrollback_empty`.

As for your first point, I'm not sure I follow about the implementation detail. This *is* the implementation, so I don't really see the issue with it knowing about the width. In the previous version of this function, it required the width to be passed in. @orhun pointed out that that was weird, and I agreed. But now the width detail is hidden in TestBackend implementaton.

I'm constructing a specific `Buffer` here so that `assert_scrollback` can be used, with all of its failure reporting goodness.

File: src/backend/test.rs [hunk 8]
```diff
@@ -140,6 +198,32 @@ impl fmt::Display for TestBackend {
     }
 }
 
+/// Append the provided cells to the bottom of a scrollback buffer. The number of cells must be a
+/// multiple of the buffer's width. If the scrollback buffer ends up larger than 65535 lines tall,
+/// then lines will be removed from the top to get it down to size.
+fn append_to_scrollback(scrollback: &mut Buffer, cells: impl IntoIterator<Item = Cell>) {
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:204]: Sure, I can do that that. I usually go the other way, personally (too many years of C code, perhaps), but I'm happy to  re-arrange it.

File: src/backend/test.rs [hunk 9]
```diff
@@ -140,6 +198,32 @@ impl fmt::Display for TestBackend {
...
+    let new_scrollback_height = scrollback.area.height as usize + lines_appended;
+    if u16::try_from(new_scrollback_height).is_ok() {
+        scrollback.area.height = new_scrollback_height as u16;
+    } else {
+        scrollback
+            .content
+            .drain(0..width * (new_scrollback_height - u16::MAX as usize));
+        scrollback.area.height = u16::MAX;
+    }
+}
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:225]: This code specifically allows the scrollback buffer to be larger than `u16::MAX` `Cell`s. That fact is actually tested in the test for overflowing scrollback.

Here I'm just trying to limit the scrollback buffer size to a certain height. The max height could have been anything that fits inside of a `u16`. I just chose `u16::MAX` because it's convenient.

In #1286 you discuss how this behavior of certain `Buffer` functions assuming the area is clamped to `u16::MAX` may not be desirable. I was trying to write code that didn't suffer from this limitation and also didn't rely on this behavior.

File: src/backend/test.rs [hunk 10]
```diff
@@ -224,16 +308,26 @@ impl Backend for TestBackend {
...
-            let rotate_by = n.saturating_sub(lines_after_cursor).min(max_y);
 
-            if rotate_by == height - 1 {
-                self.clear()?;
-            }
-
-            self.set_cursor_position(Position { x: 0, y: rotate_by })?;
-            self.clear_region(ClearType::BeforeCursor)?;
-            self.buffer.content.rotate_left((width * rotate_by).into());
+        if n > lines_after_cursor {
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:312]: Okay.

File: src/backend/test.rs [hunk 11]
```diff
@@ -676,6 +789,118 @@ mod tests {
             "eeeeeeeeee",
             "          ",
         ]);
+        backend.assert_scrollback_lines(["aaaaaaaaaa"]);
+    }
+
+    #[test]
+    fn append_multiple_lines_where_cursor_at_end_appends_more_than_height_lines() {
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:796]: It's not really that complex, but it is long.

The thing about that function (append_lines) is that it's supposed to 1) do stuff to scrollback, 2) do stuff to the screen, and 3) leave the cursor in a certain spot. That's what the test ends up doing.

I could definitely use a smaller screen height to test the same thing with fewer lines of code.

File: src/backend/test.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:225]: Ah got it. Makes sense.

Does this do the same thing as the code?:
```rust
    scrollback.content.extend(cells);
    let width = scrollback.area.width as usize;
    let new_height = (scrollback.content.len() / width).min(u16::MAX as usize);
    let keep_from = scrollback.content.len().saturating_sub(width * u16::MAX as usize);
    scrollback.content.drain(0..keep_from);
    scrollback.area.height = new_height as u16;
```

I generally avoid putting asserts inside functions like this. (as they make for difficult to debug issues for end users, and this should never be hit).

File: src/backend/test.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:796]: Github moved the code that this was attached to - it was previously attached to the append_u16_max_plus_10_lines() test

File: src/terminal/terminal.rs (hunk 3 above)
Comment (nfachan) [reply src/terminal/terminal.rs:500]: Okay. That'll requires a new impl block, but that's fine.

File: src/backend/test.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:796]: It occurred to me that you're probably referring to` append_u16_max_plus_10_lines`, not `append_multiple_lines_where_cursor_at_end_appends_more_than_height_lines`. I'll add some helper functions to it to make it more clear what's going on.

File: src/terminal/terminal.rs (hunk 5 above)
Comment (nfachan) [reply src/terminal/terminal.rs:596]: It's the latter. They sometimes go over `u16::MAX`, and sometimes go negative. The latter is especially tricky because you want to be able to write `x + y - z` or `x - y + z` and not worry about intermediate results overflowing or becoming negative.

File: src/backend/test.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:225]: Yeah, that code is equivalent. You don't need the saturating_sub, since keep_from will always be >= 0. I don't know what assert you're talking about. Are you saying you're using saturating_sub instead of vanilla - or checked_sub because the latter two effectively have asserts? If so, that's the opposite of how I usually write code, but I'll go with the flow. When I see saturating_sub, it usually indicates to me that it could saturate, whereas the other two prove to me that it won't ever. But again, I'm happy to go with the flow.

File: src/terminal/terminal.rs (hunk 2 above)
Comment (nfachan) [reply src/terminal/terminal.rs]: I rewrote the comments in the function. Please take a look and tell me what you think.

File: src/backend/test.rs [hunk 12]
```diff
@@ -224,16 +308,26 @@ impl Backend for TestBackend {
...
-
-            self.set_cursor_position(Position { x: 0, y: rotate_by })?;
-            self.clear_region(ClearType::BeforeCursor)?;
-            self.buffer.content.rotate_left((width * rotate_by).into());
+        if n > lines_after_cursor {
+            // We need to insert blank lines at the bottom and scroll the lines from the top into
+            // scrollback.
+            let scroll_by: usize = (n - lines_after_cursor).into();
+            let width: usize = self.buffer.area.width.into();
+            let to_splice = self.buffer.content.len().min(width * scroll_by);
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs]: I renamed it to `cells_to_scrollback`.

**Comment by orhun** [comment]
> It could be worth splitting the test backend changes from the terminal changes to help make it easier to review this PR. 

Sure, sounds reasonable to me.

File: src/backend/test.rs [hunk 13]
```diff
@@ -676,6 +789,118 @@ mod tests {
...
+            &scrollback_bottom,
+            &Buffer::with_lines([
+                "     65536",
+                "     65537",
+                "     65538",
+                "     65539",
+                "     65540",
+            ])
+        );
     }
```
Comment (joshka) [inline src/backend/test.rs:904]: Do these simplifications make sense:
- removed the extraneous code
- moved the scrollback size checks after the content checks (as otherwise any test failures would just report just true/false results rather than 
- added a Cell::from(char) impl
- made a few parts a bit more concise

```suggestion
    #[test]
    fn append_lines_truncates_beyond_u16_max() -> io::Result<()> {
        let mut backend = TestBackend::new(10, 5);

        // Fill the scrollback with 65535 + 10 lines
        let row_count = u16::MAX as usize + 10;
        for row in 0..=row_count {
            if row > 4 {
                backend.set_cursor_position(Position { x: 0, y: 4 })?;
                backend.append_lines(1)?;
            }
            let cells = &format!("{row:>10}").chars().map(Cell::from).collect_vec();
            let content = cells
                .iter()
                .enumerate()
                .map(|(column, cell)| (column as u16, 4.min(row) as u16, cell));
            backend.draw(content)?;
        }

        // check that the buffer contains the last 5 lines appended
        backend.assert_buffer_lines([
            "     65541",
            "     65542",
            "     65543",
            "     65544",
            "     65545",
        ]);

        // TODO: ideally this should be something like:
        //     let lines = (6..=65545).map(|row| format!("{row:>10}"));
        //     backend.assert_scrollback_lines(lines);
        // but there's some truncation happening in Buffer::with_lines that needs to be fixed

        assert_eq!(
            Buffer {
                area: Rect::new(0, 0, 10, 5),
                content: backend.scrollback.content[0..10 * 5].to_vec(),
            },
            Buffer::with_lines([
                "         6",
                "         7",
                "         8",
                "         9",
                "        10",
            ]),
            "first 5 lines of scrollback should have been truncated"
        );

        assert_eq!(
            Buffer {
                area: Rect::new(0, 0, 10, 5),
                content: backend.scrollback.content[10 * 65530..10 * 65535].to_vec(),
            },
            Buffer::with_lines([
                "     65536",
                "     65537",
                "     65538",
                "     65539",
                "     65540",
            ]),
            "last 5 lines of scrollback should have been appended"
        );

        // These checks come after the content checks as otherwise we won't see the failing content
        // when these checks fail.

        // Make sure the scrollback is the right size.
        assert_eq!(backend.scrollback.area.width, 10);
        assert_eq!(backend.scrollback.area.height, 65535);
        assert_eq!(backend.scrollback.content.len(), 10 * 65535);

        Ok(())
    }
```

This requires a small change in Cell which I think is probably reasonable:
```
impl From<char> for Cell {
    fn from(ch: char) -> Self {
        let mut cell = Cell::EMPTY;
        cell.set_char(ch);
        cell
    }
}
```

These seem like nits, but for tests I like to be able to look at them and know that the test is right at a glance rather than have to process and predict how the test works before I can known that.

**Review by joshka — COMMENTED** [review]
Added some suggestions to help clarify my previous comments.

**Comment by joshka** [comment]
> > It could be worth splitting the test backend changes from the terminal changes to help make it easier to review this PR.
> 
> Sure, sounds reasonable to me.

I spoke with @nfachan on discord and we concluded that we'd keep them in this PR rather than splitting. My concern was review speed with 400 LoC, but I think we're probably ok on this one.

File: src/backend/test.rs [hunk 14]
```diff
@@ -102,6 +112,38 @@ impl TestBackend {
...
+    /// When they are not equal, a panic occurs with a detailed error message showing the
+    /// differences between the expected and actual buffers.
+    #[allow(deprecated)]
+    #[track_caller]
+    pub fn assert_scrollback(&self, expected: &Buffer) {
+        // TODO: use assert_eq!()
+        crate::assert_buffer_eq!(&self.scrollback, expected);
+    }
+
+    /// Asserts that the `TestBackend`'s scrollback buffer is empty, but also has the given width.
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:141]: Yeah, that comment was out of date. It applied to the old version of the function where the width had to be provided.

File: src/backend/test.rs [hunk 15]
```diff
@@ -102,6 +112,38 @@ impl TestBackend {
         crate::assert_buffer_eq!(&self.buffer, expected);
     }
 
+    /// Asserts that the `TestBackend`'s scrollback buffer is equal to the expected buffer.
+    ///
+    /// This is a shortcut for `assert_eq!(self.scrollback(), &expected)`.
+    ///
+    /// # Panics
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:131]: I don't understand. I copied this from the `buffer`-based ones. Same formatting and everything.

I also don't understand the desired change. Do you really want a blank line in the middle of the comment?

File: src/backend/test.rs [hunk 16]
```diff
@@ -102,6 +112,38 @@ impl TestBackend {
...
+    #[allow(deprecated)]
+    #[track_caller]
+    pub fn assert_scrollback(&self, expected: &Buffer) {
+        // TODO: use assert_eq!()
+        crate::assert_buffer_eq!(&self.scrollback, expected);
+    }
+
+    /// Asserts that the `TestBackend`'s scrollback buffer is empty, but also has the given width.
+    ///
+    /// # Panics
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:143]: Ditto above

File: src/backend/test.rs [hunk 17]
```diff
@@ -118,6 +160,22 @@ impl TestBackend {
         self.assert_buffer(&Buffer::with_lines(expected));
     }
 
+    /// Asserts that the `TestBackend`'s scrollback buffer is equal to the expected lines.
+    ///
+    /// This is a shortcut for `assert_eq!(self.scrollback(), &Buffer::with_lines(expected))`.
+    ///
+    /// # Panics
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:179]: Ditto above

File: src/backend/test.rs [hunk 18]
```diff
@@ -263,6 +331,32 @@ impl Backend for TestBackend {
...
+    let new_scrollback_height = scrollback.area.height as usize + lines_appended;
+    if u16::try_from(new_scrollback_height).is_ok() {
+        scrollback.area.height = new_scrollback_height as u16;
+    } else {
+        scrollback
+            .content
+            .drain(0..width * (new_scrollback_height - u16::MAX as usize));
+        scrollback.area.height = u16::MAX;
+    }
+}
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:370]: Okay. I think it's a shame to remove the assert.

File: src/backend/test.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:88]: LGTM now. thanks :)

File: src/backend/test.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (orhun) [reply src/backend/test.rs:131]: Yes, this is correct. Stylelistically there should be a newline between `# Panics` and the explanation that follows.

[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:131]: But there should still be a `///` on the newline line, right? The suggested change doesn't show a `///` on the inserted line.

Should I change all of the comments, including the ones for the methods above that were pre-existing and from which I copied the format of these comments?

[the remark this answers is not in the document]
Comment (orhun) [reply src/backend/test.rs:131]: Yup, there should be a comment on the newline. It would be nice to follow the same format for all the doc comments.

**Comment by nfachan** [comment]
What do I need to do to get this PR across the finish line?

File: src/backend/test.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:132]: sorry - my bad yep that should have had an empty comment line (updated)

File: src/backend/test.rs [hunk 19]
```diff
@@ -676,6 +796,118 @@ mod tests {
...
+        ]);
+    }
+
+    #[test]
+    #[allow(deprecated)]
+    fn append_lines_truncates_beyond_u16_max() -> io::Result<()> {
+        let mut backend = TestBackend::new(10, 5);
+
+        //        backend.buffer = Buffer::empty(Rect::new(0, 0, 10, 5));
+
```
Comment (joshka) [inline src/backend/test.rs:845]:
```suggestion
```

**Review by joshka — COMMENTED** [review]
Changes look good. a couple of small things. Remaining on this is to update the title of the PR to whatever would make sense for the changelog, and provide brief summary of the changes in the main PR comment that would end up in the commit history and changelog.

PR title probably should be `fix(terminal): some short description of the bug fixed`

Something like:
```
Previously (1-2 line desciption of the bug)

This commit:
- adds a conversion constructor `Cell::from(char)`
- ...
```

These are useful because we mostly automate the changelog generation.

File: src/backend/test.rs [hunk 20]
```diff
@@ -102,11 +125,44 @@ impl TestBackend {
...
+    /// # Panics
+    ///
+    /// When they are not equal, a panic occurs with a detailed error message showing the
+    /// differences between the expected and actual buffers.
+    #[allow(deprecated)]
+    #[track_caller]
+    pub fn assert_scrollback(&self, expected: &Buffer) {
+        // TODO: use assert_eq!()
+        crate::assert_buffer_eq!(&self.scrollback, expected);
+    }
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:141]: I just copied the logic from above. I don't know why the TODO is there, or why the macro is deprecated.

The macro prints out nicer diffs, so I'm not sure why the TODO was put there or why the macro was deprecated. I figured that whoever finished that work and went to remove the TODOs and the deprecated macro would change all of the calls at once.

File: src/terminal/terminal.rs [hunk 21]
```diff
@@ -545,51 +573,121 @@ where
...
-            ..self.viewport_area
-        });
-
-        // Draw contents into buffer
+        // The approach of this function is to first render all of the lines to insert into a
+        // temporary buffer, and then to loop drawing chunks from the buffer to the screen. drawing
+        // this buffer onto the screen.
         let area = Rect {
-            x: self.viewport_area.left(),
+            x: 0,
```
[the remark this answers is not in the document]
Comment (nfachan) [reply src/terminal/terminal.rs:580]: I had the same reaction myself.

I didn't change it as I was trying to make minimal changes and keep things in the style I found.

File: src/backend/test.rs (hunk 19 above)
Comment (nfachan) [reply src/backend/test.rs:845]: Oops. I'll remove that.

File: src/terminal/terminal.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (nfachan) [reply src/terminal/terminal.rs:580]: Right, I confused myself.

I left it originally, but then changed it eventually as I ended up rewriting most of the function. Sorry for the confusion.

File: src/backend/test.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:141]: Ah got it, makes sense. The problem with the macro is that it makes the code coverage look wonky. Large amounts of actually covered code were marked as not covered for some reason. If you're making another push (no big deal if not), can you please change to assert_eq here (the buffer debug implementation should be good enough to debug failures generally).

File: src/terminal/terminal.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/terminal/terminal.rs:580]: It was only momentary. Enough to comment to save someone some investigation time, but not a problem otherwise.

File: src/backend/test.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs:141]: Feel free to resolve this comment if not changing this.

[the remark this answers is not in the document]
Comment (nfachan) [reply src/backend/test.rs:141]: Okay, I'll change it.

Why haven't the other asserting methods in this file been changed?

[the remark this answers is not in the document]
Comment (joshka) [reply src/backend/test.rs]: Likely just we haven't gotten to it.

File: src/backend/test.rs [hunk 22]
```diff
@@ -83,16 +85,37 @@ impl TestBackend {
...
 
     /// Asserts that the `TestBackend`'s buffer is equal to the expected buffer.
     ///
     /// This is a shortcut for `assert_eq!(self.buffer(), &expected)`.
     ///
     /// # Panics
+    ///
     /// When they are not equal, a panic occurs with a detailed error message showing the
     /// differences between the expected and actual buffers.
     #[allow(deprecated)]
```
Comment (joshka) [inline src/backend/test.rs:121]:
```suggestion
```

File: src/backend/test.rs [hunk 23]
```diff
@@ -676,6 +794,116 @@ mod tests {
...
+            "dddddddddd",
+            "eeeeeeeeee",
+            "          ",
+            "          ",
+            "          ",
+        ]);
+    }
+
+    #[test]
+    #[allow(deprecated)]
```
Comment (joshka) [inline src/backend/test.rs:838]:
```suggestion
```

**Review by joshka — APPROVED** [review]
I'll leave this unmerged to let @orhun take a look.

I think (not entirely sure) that the allow deprecated attributes can possibly be removed here too.

File: src/backend/test.rs (hunk 22 above)
Comment (joshka) [reply src/backend/test.rs:121]: Github moved this to the wrong line :/

**Comment by joshka** [comment]
Thanks for all your hard work on getting this over the line.

File: src/backend/test.rs (hunk 22 above)
Comment (nfachan) [reply src/backend/test.rs:121]: This one is necessary.

**Comment by nfachan** [comment]
> I'll leave this unmerged to let @orhun take a look.

I'm going to open a different PR for changes on top of these changes that address #584. I was going to wait for this to get merged in, but I don't really want to wait until after the weekend. I'd like to get eyes on it soon so I can get some feedback.

**Comment by joshka** [comment]
> > I'll leave this unmerged to let @orhun take a look.
> 
> I'm going to open a different PR for changes on top of these changes that address #584. I was going to wait for this to get merged in, but I don't really want to wait until after the weekend. I'd like to get eyes on it soon so I can get some feedback.

Sounds good. I'm going to merge this instead to make the review of the followup easier (GitHub doesn't do stacked PRs all that well unfortunately)

**Comment by orhun** [comment]
Thanks for your work on this!

**Comment by nfachan** [comment]
Thanks @orhun and @joshka for helping me get this integrated!
